src/Types/PlutusAddress.purs

```diff
@@ -0,0 +1,192 @@
+import Serialization.Address (Address) as Serialization
+import Serialization.Hash (ed25519KeyHashFromBytes, scriptHashFromBytes) as Hash
+import ToData (class ToData, toData)
+import FromData (class FromData, fromData)
+
+--------------------------------------------------------------------------------
+-- Credential
+--------------------------------------------------------------------------------
+
+data Credential
+  = PubKeyCredential PubKeyHash
+  | ScriptCredential ValidatorHash
+
+derive instance Generic Credential _
+instance Show Credential where
+  show = genericShow
+
+instance ToData Credential where
+  toData (PubKeyCredential pubKeyHash) =
+    Constr zero [ toData pubKeyHash ]
+  toData (ScriptCredential validatorHash) =
+    Constr one [ toData validatorHash ]
+
+instance FromData Credential where
+  fromData (Constr n [ pd ])
+    | n == zero = PubKeyCredential <$> fromData pd
+    | n == one = ScriptCredential <$> fromData pd
+  fromData _ = Nothing
+
+--------------------------------------------------------------------------------
+-- StakingCredential
+--------------------------------------------------------------------------------
+
+data StakingCredential
+  = StakingHash Credential
+  | StakingPtr Pointer
+
+derive instance Generic StakingCredential _
+
+instance Show StakingCredential where
+  show = genericShow
+
+instance ToData StakingCredential where
+  toData (StakingHash credential) =
+    Constr zero [ toData credential ]
+  toData (StakingPtr ptr) =
+    Constr one [ toData ptr.slot, toData ptr.txIx, toData ptr.certIx ]
+
+instance FromData StakingCredential where
+  fromData (Constr zero [ pd ]) =
+    StakingHash <$> fromData pd
+  fromData (Constr one [ slotD, txIxD, certIxD ]) =
+    StakingPtr <$> ({ slot: _, txIx: _, certIx: _} <$>
+      fromData slotD <*> fromData txIxD <*> fromData certIxD)
+  fromData _ = Nothing
+
+--------------------------------------------------------------------------------
+-- Address
+--------------------------------------------------------------------------------
+
+data PlutusAddress = Address
+  { addressCredential :: Credential
+  , addressStakingCredential :: Maybe StakingCredential
+  }
+
+derive instance Generic PlutusAddress _
+
+instance Show PlutusAddress where
+  show = genericShow
+
+instance ToData PlutusAddress where
+  toData (Address a) =
+    Constr zero [ toData a.addressCredential, toData a.addressStakingCredential ]
+
+instance FromData PlutusAddress where
+  fromData (Constr zero [ credD, stakingCredD ]) =
+    Address <$> ({ addressCredential: _, addressStakingCredential: _ } <$>
+      fromData credD <*> fromData stakingCredD)
+  fromData _ = Nothing
+
+--------------------------------------------------------------------------------
+-- IsoNativeForeign
+--------------------------------------------------------------------------------
+
+class IsoNativeForeign native foreign' | native -> foreign', foreign' -> native where
```

Comment:
Please add (a) kind signature(s), either standalone (i.e. `class IsoNativeForeign :: ...`) or on the type variables